src/main/java/org/folio/service/orders/lines/update/instance/WithHoldingOrderLineUpdateInstanceStrategy.java

```diff
@@ -62,62 +71,66 @@ protected CompletableFuture<Void> processHoldings(OrderLineUpdateInstanceHolder
-                if (ObjectUtils.notEqual(holdingId, newHoldingId)) {
-                  return updateItemsHolding(holdingId, newHoldingId, compositePoLine.getId(), requestContext);
-                } else {
-                  return CompletableFuture.completedFuture(null);
-                }
-            })
-        );
-      }
-    });
-
-    return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+      futures.add(inventoryManager.getOrCreateHoldingRecordByInstanceAndLocation(newInstanceId, location, requestContext)
+        .thenCompose(newHoldingId -> {
+          holder.addHoldingRefsToStoragePatchOrderLineRequest(holdingId, newHoldingId);
+          CompositePoLine compositePoLine = PoLineCommonUtil.convertToCompositePoLine(holder.getStoragePoLine());
+          if (ObjectUtils.notEqual(holdingId, newHoldingId)) {
+            return updateItemsHolding(holdingId, newHoldingId, compositePoLine.getId(), requestContext);
+          } else {
+            return CompletableFuture.completedFuture(null);
+          }
+        }));
+    }).thenCompose(v -> CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])));
   }
 
   private CompletableFuture<Void> createHoldingsAndUpdateItems(OrderLineUpdateInstanceHolder holder,
       String newInstanceId, RequestContext requestContext) {
-
     List<CompletableFuture<Void>> futures = new ArrayList<>();
-
-    holder.getStoragePoLine().getLocations().forEach(location -> {
+    return prepareHoldings(holder, requestContext, location -> {
       String holdingId = location.getHoldingId();
-      if (holdingId != null) {
-        futures.add(inventoryManager.createHolding(newInstanceId, location, requestContext)
-            .thenCompose(newHoldingId -> {
-                holder.addHoldingRefsToStoragePatchOrderLineRequest(holdingId, newHoldingId);
-                CompositePoLine compositePoLine = PoLineCommonUtil.convertToCompositePoLine(holder.getStoragePoLine());
-                return updateItemsHolding(holdingId, newHoldingId, compositePoLine.getId(), requestContext);
-            })
-        );
-      }
-    });
+      futures.add(inventoryManager.createHolding(newInstanceId, location, requestContext)
+        .thenCompose(newHoldingId -> {
+          holder.addHoldingRefsToStoragePatchOrderLineRequest(holdingId, newHoldingId);
+          CompositePoLine compositePoLine = PoLineCommonUtil.convertToCompositePoLine(holder.getStoragePoLine());
+          return updateItemsHolding(holdingId, newHoldingId, compositePoLine.getId(), requestContext);
+        }));
+    }).thenCompose(v -> CompletableFuture.allOf(futures.toArray(new CompletableFuture[0])));
+  }
 
-    return CompletableFuture.allOf(futures.toArray(new CompletableFuture[0]));
+  private CompletableFuture<Void> prepareHoldings(OrderLineUpdateInstanceHolder holder, RequestContext requestContext, Consumer<Location> consumer) {
```

Comment:
can you add a comment here describing how holdings are prepared?

Reply:
Renamed the method which definitely describe main purpose.